### DISCUSSION TITLE
Curious behavior of `nx.laplacian_centrality` working on directed weighted graphs

### DISCUSSION THREAD

**Discussion by ajcost** [networkx/networkx#7295] [Algorithms]
Hello NetworkX development team,

I wanted to inquire about the behavior of the Laplacian Centrality metric calculated using `nx.laplacian_centrality`. When applied to a **directed** graph, it employs the `nx.directed_laplacian_matrix` function, which produces a symmetric matrix. This approach is based on a paper by Fan Chung published in 2005.

```
.. [1] Fan Chung (2005).
       Laplacians and the Cheeger inequality for directed graphs.
       Annals of Combinatorics, 9(1), 2005
```

This paper is available here: https://sci-hub.ru/10.1007/s00026-005-0237-z. My understanding is limited in reviewing it, as this aspect of graph theory is quite specialized, and I am no expert in discrete math generally.

The code generates a symmetric matrix, crucial because the calculation of Laplacian centrality assumes the eigenvalues are real. Given the inherent asymmetry of a directed graph's adjacency matrix, which might lead to complex eigenvalues, this approach seems necessary. Yet, the rationale behind how this function creates a Laplacian matrix remains unclear to me. I'm not an expert in this area.

```python
if G.is_directed():
        lap_matrix = nx.directed_laplacian_matrix(G, nodes, weight, walk_type, alpha)
```

The implementation of Laplacian centrality is inspired by a 2012 paper by Qi, X., Fuller, E., Wu, Q., Wu, Y., and Zhang, which is cited in the code.

```
.. [1] Qi, X., Fuller, E., Wu, Q., Wu, Y., and Zhang, C.-Q. (2012).
       Laplacian centrality: A new centrality measure for weighted networks.
       Information Sciences, 194:240-253.
       https://math.wvu.edu/~cqzhang/Publication-files/my-paper/INS-2012-Laplacian-W.pdf
```

However, I find it intriguing that the 2012 paper exclusively discusses undirected weighted graphs, with no mention of directed weighted graphs. Notably, the conclusion contains an interesting remark:

> Unfortunately, Laplacian centrality also has a limitation that it cannot apply on directed networks directly. The Laplacian matrix of a directed network is not symmetric, leading to complex numbers as eigenvalues rather than real numbers. The relationship between the Laplacian energy of a directed graph and its graphical structure remains unclear. Characterizing this relationship and introducing a similar ‘‘intermediate’’ centrality method for directed graphs is also one of our future research directions.

This suggests that Laplacian centrality might not be suitable for directed graphs, or at least, the connection between a directed graph's Laplacian energy and its structure is yet to be understood. It would be surprising if the authors of the 2012 paper were not aware of Fan Chung's 2005 work if it had implications for calculating the Laplacian centrality of directed graphs.

My questions are as follows:

1. Is there additional literature that should be referenced here, explaining why the `nx.directed_laplacian_matrix` transformation is permissible and clarifying the interpretation that the authors of the 2012 Laplacian Centrality paper seem to indicate is unknown?
2. Why does NetworkX permit this calculation? Should there be a warning to users if this is considered an unconventional approach?
3. If you could clarify how the `nx.directed_laplacian_matrix` transformation aligns with the traditional concept of the Laplacian, or provide a reference clarifying this, I would greatly appreciate it.

Thank you!
Adam

**Comment by rossbar** [comment]
`laplacian_centrality` was added in #5399 - @gaborberei do you have any insight here?

**Comment by gaborberei** [comment]
Hi @ajcost, 

thanks for the detaild description this is a great insight! Honestly, I cannot answer your question as I am not a network science expert, I only wanted to help and contribute to the package and gain some experience in open source contributions. I wasn't aware the issue you described. 

I would ask @dschult about it.

**Reply by ajcost** [reply]
Hi @gaborberei , Thank you for the response! Really appreciate it and totally understand! Your implementation of Laplacian Centrality is something our team needed so we were grateful that you implemented it!

Hopefully @dschult can help us understand the interpretation of this. 

Second, I wanted to highlight a point concerning the method of calculating Laplacian energy as outlined in the paper. It seems they've opted for a non-standard approach by using the sum of the square powers of all eigenvalues, diverging from the traditional definition. Typically, this involves the sum of the absolute differences between the Laplacian eigenvalues and the average degree of the graph or, in some cases, simply the sum of the absolute values of the eigenvalues. I'm curious about the rationale behind this redefinition and would love to hear any thoughts or insights on this matter.

Nonetheless the code is correct in using the sum of squared eigenvalues.

**Reply by gaborberei** [reply]
Hi @ajcost, 

Just out of pure professional curiosity could you share some details about your work? I am just interested in why and how the laplacian centrality proved to be a better/more useful centrality measure than others. I am just interested in the specific use case, but of course I can understand if you cannot share this :)

**Reply by ajcost** [reply]
Hi @gaborberei,

Appreciate your curiosity! Unfortunately I'm not at liberty to say.

**Comment by dschult** [comment]
LR;DR.  Sorry for the late night rambling about issues bigger than all of us -- but in the end, this says: we're optimistic this is a reasonable way to handle directed graphs until evidence shows otherwise.
 
For directed graphs, NX has 3 different Laplacian Matrix functions, but only `nx.directed_laplacian_matrix` guarantees a symmetric matrix. That leads to the same computational pathway/formulas for laplacian centrality in directed and undirected graphs. It is a natural "extension" of the centrality measure used for undirected graphs. The other two laplacian matrices are defined differently: `nx.laplacian_matrix` and `nx.normalized_laplacian_matrix` extend the idea of laplacian matrix to directed graphs by using e.g. out-degree instead of `degree`, and coming to the natural extension of the `D-A` idea of laplacian matrices. Thus they are in general asymmetric.  

A few notes are in order:
- there is no reason to expect that one needs a symmetric matrix to obtain useful Laplacian Matrices for directed graphs. The Laplacian as a stand-in approximation of diffusion (on a continuous space -- which is what Laplace considered) does not provide guidance for what to do with directed underlying coordinate systems. I suspect that asymmetric matrices are the "true" simulators of directed diffusion processes -- whatever that means. Now, laplacian matrices have proven useful well beyond diffusion settings. So it is difficult to predict which extensions of the diffusion operator to directed coordinates is best in different settings.
- There is no fundamental reason to restrict centrality computations to real numbers. There is almost certainly an extension of "energy" to complex values. Some continuous diffusion-like equations, such as the Schrodinger equation have self-adjoint (symmetric) operators so result in real values for energy (eigenvalue). But many results are extended to operators that are not self-adjoint. The major technical difference with complex eigenvalues is that complex values are not an ordered set (is `3i` larger or smaller than `1`?). So we have to figure out whether the magnitude of the complex number is the correct measure of "size", or whether there is some other measure. Perhaps some applications should use the distance to a point other than the origin, or the angle in complex space between eigenvectors. The reason we restrict to real number here is that it is a natural extension of what we know. So the formulas are the same. In other words, "it's easier". We are optimistic that ideas/formulas like "Energy" can apply beyond where they have been originally defined.
- Even with real values for energy, we still don't know the best way to measure it's "size" when many objects (nodes) are combined.  "What is distance to the origin"? Historically we have used the L-2 "norm" (euclidean norm, root-sum-of-squares) to measure size. But the math behind "vector norms" shows that there are other valid "norms". I know of 3 commonly used norms: 1) root-sum-of-squares (L-2 norm because you raise it to the power 2 and then take the 2-nd root. 2) sum-of-absolute-values (L-1 norm because you raise it to the power one, take absolute value and then take the 1-th root).  3) max-norm (L-infinity norm because you raise each value to the infinite power, sum them (which means only the largest matters because it is *soo* much bigger after raising to infinity) and then take the infinity root. This 3rd is the limit of the general L-p norm when the power used is "p" and we take the limit "p to infinity". But it has a very simple outcome -- you just take the maximum of all the coordinates/values/nodes or whatever. This hopefully helps provide some context for why some people have implemented sums of absolute values, and others have implemented sums of squares.  Presumably you could just take the maximum value as well. They are all reasonable ways to measure the size of the energy. (look up L-p norm for some fun head-bending thoughts about how to produce a single number to describe a collection of values.)
-  My personal perspective is that it is naive to think of research results as either "known" results or "undiscovered yet". There are a lot of grey areas where the research communities haven't come to a consensus. Centrality measures are a good example of this. What is the "right" way to measure centrality? (There probably isn't a single right way to do it.)

So, what is the right approach to handling various measures of networks in a package like NetworkX? We prefer to restrict functions to algorithms which have some basis in published literature. But in some cases, a method has only a short history without a clear sense of whether it will pan out. Should we implement it or not? I would claim laplacian centrality (even for undirected graphs) is in that category. But once we provide an undirected method for laplacian centrality, there is a single natural extension to directed graphs that works with the existing formulation (real eigenvalues and energies). So, we're taking a bit of a leap of optimism when we provide laplacian_centrality. And we're also taking a bit of a leap of optimism to provide it for the directed case. The article said that they didn't look at the case of directed graphs. But the natural way to do that is pretty clear. It's "easy" to implement this case in our code. So we took a leap of optimism.

Anyway, that is my summary of why I think we provide laplacian centrality for directed graphs using this formulation. It has been quite a while since this code was contributed, and new methods probably have been explored. But without a paper saying that it did or did not pan out, its hard to justify changing it. Do any of you know of some results that provide some clue? I would be fine with a note in the doc_string that applying this centrality measure to directed graphs is not well studied yet. But I would argue that the same could be said of most of the centrality measures (directed or undirected). It is an active ares of research with a combinatorially large number of problems to look at. :)

**Reply by ajcost** [reply]
Hi @dschult,

Thank you for your detailed and thoughtful response. I'm particularly encouraged by the optimistic philosophy you've outlined. It seems that extending Laplacian centrality to directed graphs might be more straightforward and less fraught with challenges than I initially thought.

While searching through Google Scholar, I couldn't find any papers specifically applying this centrality measure to directed graphs, indicating a possible gap in the literature. Again I should caveat that, I am not a researcher or expert in this field, so my literature review might not be as comprehensive as someone more immersed in the domain. Your observation that this area might not be well-explored aligns with my findings.

This raises some doubts about applying a not so well-studied centrality measure to my work. But that is beside the point, given that this is a discussion on whether it should or should not be implemented in NetworkX.  The direction now seems clearer. I really value your insights on this matter.

Adam
